pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,493 @@
+	for _, eip := range eips {
+		nodeSelector, _ := metav1.LabelSelectorAsSelector(&eip.Spec.NodeSelector)
+		if nodeSelector.Matches(labels.Set(node.GetLabels())) {
+			pools.Insert(eip.Name)
+		}
+	}
+	return pools
+}
+
+func (c *Cluster) members() sets.String {
+	return sets.NewString(c.allClusterMembers()...)
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, c.externalIPPoolInformerHasSynced, c.nodeListerSynced) {
+		return
+	}
+
+	defer func() {
+		if err := c.mList.Shutdown(); err != nil {
+			klog.ErrorS(err, "Shut down cluster failed")
+		}
+		close(c.nodeEventsCh)
+	}()
+
+	n, err := c.mList.Join(c.members().List())
+	if err != nil {
+		klog.ErrorS(err, "Join cluster failed", "num", n)
+	}
+	if err := c.initConsistentHash(); err != nil {
+		klog.ErrorS(err, "Init consistentHashMap error")
+		return
+	}
+
+	for i := 0; i < defaultWorkers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	for {
+		select {
+		case <-stopCh:
+			return
+		case nodeEvent := <-c.nodeEventsCh:
+			c.handleClusterNodeEvents(&nodeEvent)
+		}
+	}
+}
+
+func (c *Cluster) initConsistentHash() error {
+	node, err := c.nodeLister.Get(c.nodeName)
+	if err != nil {
+		return err
+	}
+	eips := c.filterEIPsFromNodeLabels(node)
+	c.enqueueEIPs(eips)
+	return nil
+}
+
+func (c *Cluster) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Cluster) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect strings (externalIPPool name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncConsistenHash(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Syncing failed, requeue", "eip", key)
+	}
+	return true
+}
+
+func (c *Cluster) syncConsistenHash(eipName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).InfoS("Finished syncing consistentHash", "eip", eipName, "durationTime", time.Since(startTime))
+	}()
+
+	c.consistentHashRWMutex.Lock()
+	defer c.consistentHashRWMutex.Unlock()
+	eip, err := c.externalIPPoolLister.Get(eipName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			delete(c.consistentHashMap, eipName)
+			return nil
+		}
+		return err
+	}
+
+	if err := c.syncConsistentHashByEIP(eip); err != nil {
```

Comment:
`syncConsistentHashByEIP` is the only logic of `syncConsistenHash` and it's not shared by other methods, how about just putting its code here?

Reply:
done